Sources/NIOPCAP/PCAP.swift

```diff
@@ -0,0 +1,469 @@
+        case v6(src: SocketAddress.IPv6Address, dst: SocketAddress.IPv6Address)
+
+        public var srcPort: UInt16 {
+            switch self {
+            case .v4(src: let src, dst: _):
+                return UInt16(bigEndian: src.address.sin_port)
+            case .v6(src: let src, dst: _):
+                return UInt16(bigEndian: src.address.sin6_port)
+            }
+        }
+
+        public var dstPort: UInt16 {
+            switch self {
+            case .v4(src: _, dst: let dst):
+                return UInt16(bigEndian: dst.address.sin_port)
+            case .v6(src: _, dst: let dst):
+                return UInt16(bigEndian: dst.address.sin6_port)
+            }
+        }
+    }
+
+    public var payloadLength: Int
+    public var addresses: AddressTuple
+    public var time: timeval
+    public var tcp: TCPHeader
+
+    public init(payloadLength: Int, addresses: AddressTuple, time: timeval, tcp: TCPHeader) {
+        self.payloadLength = payloadLength
+        self.addresses = addresses
+        self.time = time
+        self.tcp = tcp
+
+        assert(addresses.srcPort == Int(tcp.srcPort))
+        assert(addresses.dstPort == Int(tcp.dstPort))
+        assert(tcp.ackNumber == nil ? !tcp.flags.contains([.ack]) : tcp.flags.contains([.ack]))
+    }
+
+    public init(payloadLength: Int, src: SocketAddress, dst: SocketAddress, tcp: TCPHeader) throws {
+        let addressTuple: AddressTuple
+        switch (src, dst) {
+        case (.v4(let src), .v4(let dst)):
+            addressTuple = .v4(src: src, dst: dst)
+        case (.v6(let src), .v6(let dst)):
+            addressTuple = .v6(src: src, dst: dst)
+        default:
+            throw Error.incompatibleAddressPair(src, dst)
+        }
+        self = .init(payloadLength: payloadLength, addresses: addressTuple, tcp: tcp)
+    }
+
+    public init(payloadLength: Int, addresses: AddressTuple, tcp: TCPHeader) {
+        var tv = timeval()
+        gettimeofday(&tv, nil)
+        self = .init(payloadLength: payloadLength, addresses: addresses, time: tv, tcp: tcp)
+    }
+}
+
+public struct PCAPRecord {
+    public var time: timeval
+    public var header: PCAPRecordHeader
+    public var pcapProtocolID: UInt32
+    public var payload: ByteBuffer
+}
+
+public struct TCPIPv4Packet {
+    public var src: in_addr
+    public var dst: in_addr
+    public var wholeIPPacketLength: Int
+    public var tcpHeader: TCPHeader
+    public var rawTCPOptions: ByteBuffer
+    public var tcpPayload: ByteBuffer
+}
+
+public struct TCPIPv6Packet {
+    public var src: in6_addr
+    public var dst: in6_addr
+    public var payloadLength: Int
+    public var tcpHeader: TCPHeader
+    public var tcpPayload: ByteBuffer
+}
+
+extension ByteBuffer {
+    // read & parse a TCP packet, containing everything belonging to it (including payload)
+    public mutating func readTCPHeader() throws -> TCPHeader? {
+        struct ParsingError: Error {
```

Comment:
Should this error be declared at a higher scope?

Reply:
thanks, done